# CLEANING EBAY CAR SALES DATA

This project uses the data from eBay Kleinanzeigen, a section of the German eBay website. The aim of this project is to clean and analyze the car lisitings data using pandas.

### IMPORT PANDAS AND NUMPY LIBRARIES

In [1]:
import pandas as pd
import numpy as np

### READ FILE INTO DATAFRAME

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

### QUICK EXPLORATION OF DATA

In [3]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

All columns (15 columns) are string objects except year of Registration, powerPS, month of Registration, number of pictures and postal code.

Vehicle type, model, fuel type and notRepaired Damaged columns have null values

In [4]:
autos.shape

(50000, 20)

There are 50,000 rows and 20 columns in the data

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.tail()

dateCrawled                                               name  \
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

      brand notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
49995  audi              nein  2016-03-27 00:00:00             0       82131   
49996  opel              nein  2016-03-28 00:00:00             0       44807   
49997  fiat              nein  2016-04-02 00:00:00             0       73430   
49998  audi              nein  2016-03-08 00:00:00             0       35683   
49999  opel              nein  2016-03-13 00:00:00             0       45897   

                  lastSeen  
49995  2016-04-01 13:47:40  
49996  2016-04-02 14:18:02  
49997  2016-04-04 11:47:27  
49998  2016-04-05 16:45:07  
49999  2016-04-06 21:18:48

The data set uses integer based label

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

### Re-adjust column names
From the above, the columns uses camel case notoation instead of python's snake case. The columns are changed to snake case notat

In [8]:
autos = autos.rename(columns = {'dateCrawled': 'date_crawled',
                 'abtest': 'ab_test',
                 'vehicleType': 'vehicle_type',
                 'offerType': 'offer_type',
                 'yearOfRegistration': 'registration_year',
                 'gearbox': 'gear_box',
                 'powerPS': 'power_ps',
                 'monthOfRegistration': 'registration_month',
                 'fuelType': 'fuel_type',
                 'notRepairedDamage' : 'unrepaired_damage',
                 'dateCreated': 'ad_created',
                 'nrOfPictures': 'num_photos',
                 'postalCode': 'postal_code',
                 'lastSeen': 'last_seen'})
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')


All columns have now been changed to snake case notation. 

### INIITIAL EXPLORATION AND CLEANING

To determine errors and what cleaning to be done, the dataframe must be explored. Common problems to look for include:

1. Test columns where all or almost all values are the same - Duplicate observations are not needed for analysis.

2. Numeric data stored as text.



In [9]:
#Get descriptive statistics for all columns
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

The seller, offertype and num_photos columns are candidates columns to be dropped because they contain mostly only one value.

In [10]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [11]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [12]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

In [13]:
autos['ab_test'].value_counts()

test       25756
control    24244
Name: ab_test, dtype: int64

In [14]:
autos = autos.drop(['seller', 'offer_type', 'num_photos'], axis = 1)
len(autos.columns)

17

Next step is to check for numeric values stored as text

In [15]:
autos.describe(include = 'all')

date_crawled         name  price ab_test vehicle_type  \
count                 50000        50000  50000   50000        44905   
unique                48213        38754   2357       2            8   
top     2016-04-04 16:40:33  Ford_Fiesta     $0    test    limousine   
freq                      3           78   1421   25756        12859   
mean                    NaN          NaN    NaN     NaN          NaN   
std                     NaN          NaN    NaN     NaN          NaN   
min                     NaN          NaN    NaN     NaN          NaN   
25%                     NaN          NaN    NaN     NaN          NaN   
50%                     NaN          NaN    NaN     NaN          NaN   
75%                     NaN          NaN    NaN     NaN          NaN   
max                     NaN          NaN    NaN     NaN          NaN   

        registration_year gear_box      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

The price and odomoter columns are the two candidates for this,
from the descriptive statistics above. The value for price and odometer are numeric but contain strings i.e. $ sign and kms.

This is further investigated below, to assure the assumption is correct

In [16]:
print(autos['price'].head(2))
autos['odometer'].head(2)

0    $5,000
1    $8,500
Name: price, dtype: object


0    150,000km
1    150,000km
Name: odometer, dtype: object

Price and odometer are both confirmed to be numeric values stored as strings, as well as containing string attributes. The curse of action is to strip the columns off their string elements and convert them to their appropriate datatype compatible for data analysis.

In [17]:
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '')
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].str.replace('km', '')

In [18]:
print(autos['price'].head(2))
autos['odometer'].head(2)

0    5000
1    8500
Name: price, dtype: object


0    150000
1    150000
Name: odometer, dtype: object

The datatypes are still objects datatypes, they will both be changed to int.

In [19]:
autos['price'] = autos['price'].astype(int)
autos['odometer'] = autos['odometer'].astype(int)

In [20]:
print(autos['price'].head(2))
autos['odometer'].head(2)

0    5000
1    8500
Name: price, dtype: int64


0    150000
1    150000
Name: odometer, dtype: int64

The last step will be to change the column name for odometer, specifiying the data is stored in km, for better understanding.

In [21]:
autos.rename({'odometer': 'odometer_km'}, axis = 1, inplace = True)
print(autos.columns)

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gear_box', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


### Further analysis of odometer and price columns
To ensure the price and odometer columns are free of outliers, an analyzation of their basic statistics will be carried out, in search of unrealistically high and low values.

In [22]:
#Check the number of unique values for price column
print('Price unique values : ', autos['price'].unique().shape)
#Get the basic statisitcs of price column
print()
print('Statistics for price column')
print(autos['price'].describe())
print()
#view the counts of each unique values
autos['price'].value_counts()

Price unique values :  (2357,)

Statistics for price column
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64



0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price, Length: 2357, dtype: int64

There are 2357 unique values in the price column. From the unique value counts, it can be seen that there are 1421 price of $0 (about 60 percent of the unique values) this does not procide any need for analysis, and thus will be removed.

In [23]:
autos = autos[autos['price'] > 0]

In [24]:
print(autos['price'].value_counts().head())
print()
print(autos['price'].describe())
print()
print('Price unique values : ', autos['price'].unique().shape)

500     781
1500    734
2500    643
1000    639
1200    639
Name: price, dtype: int64

count    4.857900e+04
mean     1.012788e+04
std      4.880873e+05
min      1.000000e+00
25%      1.200000e+03
50%      3.000000e+03
75%      7.490000e+03
max      1.000000e+08
Name: price, dtype: float64

Price unique values :  (2356,)


In [25]:
print('Odometer unique values : ',\
      autos['odometer_km'].unique().shape)
print('Statistics for odometer column')
autos['odometer_km'].describe()

Odometer unique values :  (13,)
Statistics for odometer column


count     48579.000000
mean     125766.174685
std       39792.936148
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [26]:
autos['odometer_km'].value_counts()

150000    31422
125000     5058
100000     2116
90000      1734
80000      1415
70000      1217
60000      1155
50000      1014
5000        837
40000       816
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

The values in the odometer column, seems pretty valid without any outliers. Thus, will remain the same

### Exploring the Date Columns

In [27]:
autos.head(1)

date_crawled                              name  price  ab_test  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD   5000  control   

  vehicle_type  registration_year gear_box  power_ps   model  odometer_km  \
0          bus               2004  manuell       158  andere       150000   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

There are 5 main columns representing date values in the dataset:

1. date_crawled

2. last_seen

3. ad_created

4. registration_year

5. registration_month

The datatypes of each column needs to be explored, to ensure it is appropriate

In [28]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48579 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled          48579 non-null object
name                  48579 non-null object
price                 48579 non-null int64
ab_test               48579 non-null object
vehicle_type          43990 non-null object
registration_year     48579 non-null int64
gear_box              46232 non-null object
power_ps              48579 non-null int64
model                 46116 non-null object
odometer_km           48579 non-null int64
registration_month    48579 non-null int64
fuel_type             44544 non-null object
brand                 48579 non-null object
unrepaired_damage     39472 non-null object
ad_created            48579 non-null object
postal_code           48579 non-null int64
last_seen             48579 non-null object
dtypes: int64(6), object(11)
memory usage: 6.7+ MB
None


Columns registration_month and registration_year are integers, whereas the other date columns are string objects, which have to be converted numerically.

In [29]:
print('Get a basic understanding of the format of the date columns')
autos[['date_crawled', 'ad_created', 'last_seen']] [0:2]

Get a basic understanding of the format of the date columns


date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08

To enable us, analyze the data further, the time information on these columns will be removed. Since the columns are strings, the time starts at index 11. Thus, we slice the data from start to index 10.

In [30]:
autos['date_crawled'] = autos['date_crawled'].str[:10]
autos['ad_created'] = autos['ad_created'].str[:10]
autos['last_seen'] = autos['last_seen'].str[:10]
autos[['date_crawled', 'ad_created', 'last_seen']] [0:2]

date_crawled  ad_created   last_seen
0   2016-03-26  2016-03-26  2016-04-06
1   2016-04-04  2016-04-04  2016-04-06

In [31]:
print(autos['date_crawled'].value_counts\
      (dropna = False).sort_values())

2016-04-07      68
2016-04-06     154
2016-03-18     627
2016-04-05     636
2016-03-06     682
2016-03-13     761
2016-03-05    1230
2016-03-24    1425
2016-03-16    1438
2016-03-27    1510
2016-03-25    1535
2016-03-17    1537
2016-03-31    1547
2016-03-10    1563
2016-03-26    1564
2016-03-23    1565
2016-03-11    1582
2016-03-22    1604
2016-03-09    1608
2016-03-08    1619
2016-03-30    1636
2016-04-01    1636
2016-03-29    1658
2016-03-15    1665
2016-03-19    1689
2016-03-28    1693
2016-04-02    1723
2016-03-07    1749
2016-04-04    1774
2016-03-14    1775
2016-03-12    1794
2016-03-21    1817
2016-03-20    1840
2016-04-03    1875
Name: date_crawled, dtype: int64


In [32]:
print(autos['ad_created'].value_counts\
      (dropna = False).sort_values())

2015-12-05       1
2016-01-22       1
2015-09-09       1
2016-01-03       1
2016-02-08       1
2016-02-22       1
2016-01-14       1
2015-12-30       1
2016-02-16       1
2016-02-09       1
2016-01-16       1
2016-02-11       1
2015-08-10       1
2016-01-07       1
2016-01-29       1
2016-02-17       1
2015-11-10       1
2016-01-13       1
2016-02-07       1
2015-06-11       1
2016-02-01       1
2016-02-14       2
2016-02-24       2
2016-02-20       2
2016-02-12       2
2016-02-02       2
2016-02-05       2
2016-02-18       2
2016-01-10       2
2016-02-26       2
              ... 
2016-03-06     744
2016-03-13     826
2016-03-05    1112
2016-03-24    1422
2016-03-16    1463
2016-03-27    1505
2016-03-17    1520
2016-03-25    1542
2016-03-10    1549
2016-03-31    1549
2016-03-23    1557
2016-03-26    1567
2016-03-22    1595
2016-03-11    1598
2016-03-09    1611
2016-03-08    1620
2016-03-30    1627
2016-03-19    1636
2016-04-01    1636
2016-03-15    1652
2016-03-29    1655
2016-03-07  

In [33]:
print(autos['last_seen'].value_counts\
      (dropna = False).sort_values())

2016-03-05       52
2016-03-06      210
2016-03-07      262
2016-03-18      357
2016-03-08      361
2016-03-13      432
2016-03-09      467
2016-03-10      518
2016-03-11      601
2016-03-14      612
2016-03-27      760
2016-03-19      769
2016-03-15      771
2016-03-16      799
2016-03-26      816
2016-03-23      900
2016-03-25      933
2016-03-24      960
2016-03-20     1003
2016-03-21     1003
2016-03-28     1014
2016-03-22     1038
2016-03-29     1086
2016-04-01     1107
2016-03-31     1156
2016-03-12     1156
2016-04-04     1189
2016-03-30     1203
2016-04-02     1211
2016-04-03     1224
2016-03-17     1364
2016-04-05     6061
2016-04-07     6408
2016-04-06    10776
Name: last_seen, dtype: int64


### EXPLORING REGISTRATION YEAR

In [34]:
autos['registration_year'].describe()

count    48579.000000
mean      2004.753000
std         88.631663
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

From the above, certain erros are available in this column.

1. Count of the column is 48579 meaning there are about 1421 missing values.

2. The minimum year is 1000 - long before cars were invented

3. The maximum year is 9999 - we are in 2019.

The second and third observation shows that outliers are present in the dataset.

In [35]:
(~autos['registration_year'].between(1900,2020)).sum() \
 / autos.shape[0]

0.0004117005290351798

From a realstic point of view, any car registration that comes before 1900 and after 2019, will be considered as an outlier. Because cars were invented at the late 1800s and a car cannot be registered before it's availability.

Given, the small percentage of this values, the focus will be to remove the outlying values.

In [36]:
autos = autos[autos['registration_year'].between(1900,2019)]
autos["registration_year"].value_counts().sort_values(ascending = False).head(10
                                                                             )

2000    3156
2005    2936
1999    2899
2004    2703
2003    2699
2006    2671
2001    2637
2002    2486
1998    2363
2007    2277
Name: registration_year, dtype: int64

It appears most car registration occured during the 2000s.

In [37]:
autos['registration_year'].describe()

count    48559.000000
mean      2003.459462
std          7.571165
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

As we can see the new minimum year is 1910, and the maximum is 2019

### EXPLORING PRICE BY BRAND

In [38]:
autos['brand'].value_counts()

volkswagen        10334
bmw                5275
opel               5275
mercedes_benz      4651
audi               4168
ford               3384
renault            2325
peugeot            1430
fiat               1263
seat                919
skoda               780
nissan              741
mazda               739
smart               694
citroen             686
toyota              611
hyundai             483
sonstige_autos      471
volvo               440
mini                418
mitsubishi          397
honda               388
kia                 345
alfa_romeo          321
porsche             287
suzuki              286
chevrolet           275
chrysler            169
dacia               129
daihatsu            122
jeep                107
subaru              102
land_rover           99
saab                 79
daewoo               76
jaguar               74
trabant              67
rover                65
lancia               55
lada                 29
Name: brand, dtype: int64

Volkswagen has the highest number of car listing.There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings

In [39]:
brand_counts = autos['brand'].value_counts(normalize = True)
popular_cars = brand_counts[brand_counts > .05].index
print(popular_cars)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


The next step is to see how the prices are distributed among the popular cars.

In [43]:
brand_mean_price = {}
for brand in popular_cars:
    brand_in_data = autos[autos['brand'] == brand]
    mean_price = brand_in_data['price'].mean()
    brand_mean_price[brand] = int(mean_price)
    
brand_mean_price

{'audi': 9212,
 'bmw': 8493,
 'ford': 7305,
 'mercedes_benz': 30025,
 'opel': 5281,
 'volkswagen': 6600}

Of the top 6 car brands, Mercedes Benz, Audi and BMW are the most expensive respectively.
1. Mercedes Benz
2. Audi
3. BMW
4. Ford
5. Volkswagen
6. Opel

### EXPLORING MILEAGE
After knowing the top 6 car brands, we will like to see if the mileage of these cars have any visible link with the mean price

In [49]:
mean_mileage_dict = {}
for brand in popular_cars:
    brand_in_data = autos[autos['brand'] == brand]
    mean_mileage = brand_in_data['odometer_km'].mean()
    mean_mileage_dict[brand] = int(mean_mileage)
mean_mileage_dict

{'audi': 129492,
 'bmw': 132686,
 'ford': 124327,
 'mercedes_benz': 130852,
 'opel': 129455,
 'volkswagen': 128961}

The descending order of the mileage values are:

1. BMW

2. Mercedes benz

3. Audi

4. Opel

5. Volkswagen

6. Ford


In [50]:
bpp = pd.Series(brand_mean_price)
print(bpp)

audi              9212
bmw               8493
ford              7305
mercedes_benz    30025
opel              5281
volkswagen        6600
dtype: int64


In [51]:
bmp = pd.Series(mean_mileage_dict)
print(bmp)

audi             129492
bmw              132686
ford             124327
mercedes_benz    130852
opel             129455
volkswagen       128961
dtype: int64


In [65]:
price_vs_mileage = pd.DataFrame(bmp, columns = ['mean_mileage'])
price_vs_mileage
price_vs_mileage ['mean_price'] = bpp
price_vs_mileage

mean_mileage  mean_price
audi                 129492        9212
bmw                  132686        8493
ford                 124327        7305
mercedes_benz        130852       30025
opel                 129455        5281
volkswagen           128961        6600

There is a trend for the more expensive cars having a higher mileage. Like Mercedes Benz being the most expensive and second highest mileage.  And volkswagen being second to the least expensive with second to the least mileage.

However the correlation between both variables are not strong.